src/java/org/apache/cassandra/locator/TokenMetadata.java

```diff
@@ -174,6 +183,7 @@ public void updateNormalTokens(Multimap<InetAddress, Token> endpointTokens)
         if (endpointTokens.isEmpty())
             return;
 
+        publicLock.readLock().lock();
         lock.writeLock().lock();
```

Comment:
Should we always acquire `lock` before `publicLock` to ensure consistent ordering and no deadlocks? Do we have concurrency tests to hammer these and verify correctness and liveness guarantees?

Reply:
Good flag -- yes, my plan is after we smoke tested that the change actually works, my next step is to create a couple of tests to ensure we cannot deadlock.